File: haskell/ac05.hs
import Data.Char
import Data.List
import Data.Maybe

cut p list =
  let (hd, (_:tl)) = break p list
  in (hd, tl)

split p list =
  let (hd, tl) = break p list
  in case tl of
    [] -> [hd]
    _:tl -> hd:split p tl

data Line = Line {
  dst :: Int,
  src :: Int,
  len :: Int
}

parseLine line =
  let (dst_str, tail) = cut (== ' ') line
      (src_str, len_str) = cut (== ' ') tail
  in Line { dst = read dst_str, src = read src_str, len = read len_str }

parseMaps lines =
  case lines of
    [] -> []
    _emptyline:_title:value_lines ->
      let (values_str, tail) = break (== []) value_lines
          values = map parseLine values_str
      in values:parseMaps tail

parseInput (seeds_str:maps_lines) =
  let seeds = map read (split (== ' ') (drop (length "seeds: ") seeds_str))
      maps = parseMaps maps_lines
  in (seeds, maps)

singleImageOfLine value line =
  if value >= src line && value < src line + len line then
    Just (dst line + value - src line)
  else
    Nothing

singleImageOfMap value m =
  case catMaybes (map (singleImageOfLine value) m) of
    image:_ -> image
    [] -> value

data Interval = Interval {
  isrc :: Int,
  ilen :: Int
}

peekInterval list =
  case list of
    src:len:tail -> Just (Interval { isrc = src, ilen = len }, tail)
    _ -> Nothing

intervalBetween isrc end =
  if isrc < end then
    [Interval { isrc = isrc, ilen = end - isrc }]
  else
    []

intervalImageOfLine line (remaining_intervals, new_intervals) interval =
  let line_end = src line + len line
      interval_end = isrc interval + ilen interval
  in if line_end <= isrc interval || interval_end <= src line then
       (interval:remaining_intervals, new_intervals)
     else
       let begin = max (src line) (isrc interval)
           end = min line_end interval_end
       in (intervalBetween (isrc interval) begin ++
           intervalBetween end interval_end ++
           remaining_intervals,
           Interval { isrc = dst line + begin - src line, ilen = end - begin }:new_intervals)

intervalImagesOfLine (intervals, new_intervals) line =
  foldl (intervalImageOfLine line) ([], new_intervals) intervals

intervalImagesOfMap intervals map =
  let (remaining_intervals, new_intervals) = foldl intervalImagesOfLine (intervals, []) map
  in remaining_intervals ++ new_intervals

main = do
  input <- getContents
  let (seeds, maps) = parseInput (lines input)
  let values = map (\value -> foldl singleImageOfMap value maps) seeds
  let result_part1 = foldr1 min values
  putStrLn ("Part 1: " ++ show result_part1)
  let seed_intervals = unfoldr peekInterval seeds
  let intervals = foldl intervalImagesOfMap seed_intervals maps
  let result_part2 = foldr1 min (map isrc intervals)
  putStrLn ("Part 2: " ++ show result_part2)

File: rust/src/bin/ac05.rs
struct Line {
    dst: u64,
    src: u64,
    len: u64,
}

type Map = Vec<Line>;

struct Range {
    src: u64,
    len: u64,
}

fn main() {
    let mut lines = std::io::stdin().lines();
    let seeds: Vec<u64> =
	lines.next().unwrap().unwrap().as_str().strip_prefix("seeds: ").unwrap()
	.split(" ").map(|s| s.parse().unwrap()).collect();
    let mut lines = lines.skip(1);
    let mut maps = Vec::new();
    while let Some(_) = lines.next() {
	let mut map = Map::new();
	while let Some(line) = lines.next() {
	    let line = line.unwrap();
	    if line == "" {
		break;
	    }
	    let mut values =
                line.as_str().split(" ").map(|s| s.parse().unwrap());
	    let dst = values.next().unwrap();
	    let src = values.next().unwrap();
	    let len = values.next().unwrap();
	    map.push(Line { dst, src, len });
	}
	maps.push(map);
    }
    let mut values = seeds.clone();
    for map in &maps {
	let mut next_values = Vec::new();
	for line in map {
	    values.retain(
		|&value|
		if value >= line.src && value < line.src + line.len {
		    next_values.push(line.dst + value - line.src);
		    false
		}
		else {
		    true
		}
	    )
	}
	next_values.append(&mut values);
	values = next_values;
    }
    let result_part1 = values.iter().min().unwrap();
    let mut ranges = Vec::new();
    let mut seeds_iter = seeds.iter();
    while let Some(&src) = seeds_iter.next() {
	ranges.push(Range { src, len: *seeds_iter.next().unwrap() });
    }
    for map in maps.iter() {
	let mut next_ranges = Vec::new();
	for line in map {
	    let mut remaining_ranges = Vec::new();
	    ranges.retain(
		|range|
		if
		    range.src + range.len < line.src ||
		    line.src + line.len < range.src
		{
		    true
		}
		else {
		    let range_end = range.src + range.len;
		    let src = std::cmp::max(range.src, line.src);
		    let end = std::cmp::min(range_end, line.src + line.len);
		    next_ranges.push(
			Range { src: src + line.dst - line.src, len: end - src });
		    if src > range.src {
			remaining_ranges.push(
			    Range { src: range.src, len: src - range.src });
		    }
		    if end < range_end {
			remaining_ranges.push(
			    Range { src: end, len: range_end - end });
		    }
		    false
		}
	    );
	    ranges.append(&mut remaining_ranges);
	}
	next_ranges.append(&mut ranges);
	ranges = next_ranges;
    }
    let result_part2 = ranges.iter().map(|range| range.src).min().unwrap();
    println!("Part 1: {result_part1}");
    println!("Part 2: {result_part2}");
}
